pyccel/codegen/printing/cwrappercode.py

```diff
@@ -0,0 +1,351 @@
+# coding: utf-8
+# pylint: disable=R0201
+from collections import OrderedDict
+
+from pyccel.codegen.printing.ccode import CCodePrinter
+
+from pyccel.ast.numbers   import BooleanTrue
+from pyccel.ast.builtins import Bool
+
+from pyccel.ast.core import Variable, ValuedVariable, Assign, AliasAssign, FunctionDef
+from pyccel.ast.core import If, Nil, Return, FunctionCall, PyccelNot, Symbol, Constant
+from pyccel.ast.core import create_incremented_string, Declare, SeparatorComment
+from pyccel.ast.core import IfTernaryOperator
+
+from pyccel.ast.datatypes import NativeInteger, NativeBool, NativeComplex, NativeReal
+
+from pyccel.ast.cwrapper import PyccelPyObject, PyArg_ParseTupleNode, PyBuildValueNode
+from pyccel.ast.cwrapper import PyArgKeywords, FuncCall
+from pyccel.ast.cwrapper import Py_True, Py_False
+
+from pyccel.ast.type_inference import str_dtype
+
+from pyccel.errors.errors import Errors
+from pyccel.errors.messages import *
+
+errors = Errors()
+
+__all__ = ["CWrapperCodePrinter", "cwrappercode"]
+
+dtype_registry = {('pyobject', 0) : 'PyObject'}
+
+class CWrapperCodePrinter(CCodePrinter):
+    def __init__(self, parser, settings={}):
+        CCodePrinter.__init__(self, parser,settings)
+        self._cast_functions_dict = OrderedDict()
+        self._function_wrapper_names = dict()
+        self._global_names = set()
+
+    def get_new_name(self, used_names, requested_name):
+        if requested_name not in used_names:
+            used_names.add(requested_name)
+            return requested_name
+        else:
+            incremented_name, _ = create_incremented_string(used_names, prefix=requested_name)
+            return incremented_name
+
+    def find_in_dtype_registry(self, dtype, prec):
+        try :
+            return dtype_registry[(dtype, prec)]
+        except KeyError:
+            return CCodePrinter.find_in_dtype_registry(self, dtype, prec)
+
+    def get_cast_function(self, used_names, cast_type, from_variable, to_variable):
```

Comment:
The set `used_names` contains all the names used in the current function. Here we are creating a new function. Variables local to this function cannot have a name collision with variables local to another function. Therefore `used_names` should not be passed. We should use the set `self._global_names` to avoid collisions with names in the global namespace

Reply:
If the variable passed in the cast function have the same name as the element of this function this will create a problem 
```
@types('complex')
def func(x):
    real_part = x
    return real_part
```
the C code will be 
```
PyObject *complex_to_pycomplex(double _Complex real_part)
{
    double real_part;
    etc...........
```

Reply:
You mean if the argument has the same name as an element of the function. We can do:
```
PyObject *complex_to_pycomplex(double _Complex real_arg)
{
    double real_part;
    etc...........
}

double _Complex real_part;
complex_to_pycomplex(real_part);
```
The cast function should always be the same unless the user's function is called `complex_to_pycomplex` I think.

I would be inclined to define all the cast functions in `ast/cwrapper.py` and then just use `get_cast_function` to return the right one

Reply:
That way it is clear that the definition is separate from the call, even if that means that all the possible conversions exist as objects in the python memory (this is not a huge problem, python is memory hungry anyway. We tend to prefer readability where possible)